File: baseLib/src/androidTest/java/me/ycdev/android/lib/common/test/ObjectLeakCheckerTest.java
package me.ycdev.android.lib.common.test;

import android.support.test.runner.AndroidJUnit4;

import org.junit.Test;
import org.junit.runner.RunWith;

import static org.junit.Assert.assertEquals;

@RunWith(AndroidJUnit4.class)
public class ObjectLeakCheckerTest {
    private static class Dummy {}

    private static Dummy sHolder;

    @Test
    public void testNoLeak() {
        ObjectLeakChecker.ObjectOperator<Dummy> operator = new ObjectLeakChecker.ObjectOperator<Dummy>() {
            @Override
            public Dummy createObject() {
                return new Dummy();
            }

            @Override
            public void operate(Dummy obj) {
                // nothing to do
            }
        };
        ObjectLeakChecker<Dummy> checker = new ObjectLeakChecker<>(operator);
        checker.prepareForGc();
        checker.waitGcDone();
        assertEquals(0, checker.getLeakedObjectCount());
    }

    @Test
    public void testLeak() {
        ObjectLeakChecker.ObjectOperator<Dummy> operator = new ObjectLeakChecker.ObjectOperator<Dummy>() {
            @Override
            public Dummy createObject() {
                return new Dummy();
            }

            @Override
            public void operate(Dummy obj) {
                sHolder = obj;
            }
        };
        ObjectLeakChecker<Dummy> checker = new ObjectLeakChecker<>(operator);
        checker.prepareForGc();
        checker.waitGcDone();
        assertEquals(1, checker.getLeakedObjectCount());
    }
}

File: testLib/src/androidTest/java/me/ycdev/android/lib/test/ObjectLeakCheckerTest.kt
package me.ycdev.android.lib.test

import com.google.common.truth.Truth.assertThat
import org.junit.Test

class ObjectLeakCheckerTest {
    private class Dummy

    @Test
    fun testNoLeak() {
        val operator = object : ObjectLeakChecker.ObjectOperator<Dummy> {
            override fun createObject(): Dummy {
                return Dummy()
            }

            override fun operate(obj: Dummy) {
                // nothing to do
            }
        }
        val checker = ObjectLeakChecker(operator)
        checker.prepareForGc()
        checker.waitGcDone()
        assertThat(checker.leakedObjectCount).isEqualTo(0)
    }

    @Test
    fun testLeak() {
        val operator = object : ObjectLeakChecker.ObjectOperator<Dummy> {
            override fun createObject(): Dummy {
                return Dummy()
            }

            override fun operate(obj: Dummy) {
                holder = obj
            }
        }
        val checker = ObjectLeakChecker(operator)
        checker.prepareForGc()
        checker.waitGcDone()
        assertThat(checker.leakedObjectCount).isEqualTo(1)
    }

    companion object {

        private var holder: Dummy? = null
    }
}
